lib/rucio/client/downloadclient.py

```diff
@@ -646,6 +559,387 @@ def download(self, dids, rse=None, archive=None, protocol=None, pfn=None, nrando
+        (This function is meant to be used as class internal only)
+
+        :param items: list of dictionaries containing the items to prepare
+
+        :returns: list of dictionaries, one dict for each file to download
+
+        :raises InputValidationError: if the given input is not valid or incomplete
+        """
+        logger = self.logger
+
+        logger.info('Processing %d item(s) for input' % len(items))
+        resolved_items = []
+        # resolve input: extend rse expression, resolve wildcards, get did type
+        for item in items:
+            did_str = item.get('did')
+            if not did_str:
+                raise InputValidationError('The key did is mandatory')
+
+            logger.debug('Processing item %s' % did_str)
+
+            new_item = copy.deepcopy(item)
+
+            # extend RSE expression to exclude tape RSEs for non-admin accounts
+            if not self.is_admin:
+                rse = new_item.get('rse')
+                new_item['rse'] = 'istape=False' if not rse else '(%s)&istape=False' % rse
+                logger.debug('RSE-Expression: %s' % new_item['rse'])
+
+            # resolve any wildcards in the input dids
+            did_scope, did_name = self._split_did_str(did_str)
+            logger.debug('Splitted DID: %s:%s' % (did_scope, did_name))
+            new_item['scope'] = did_scope
+            if '*' in did_name:
+                logger.debug('Resolving wildcarded DID %s' % did_str)
+                for dids in self.client.list_dids(did_scope, filters={'name': did_name}, type='all', long=True):
+                    logger.debug('%s - %s:%s' % (dids['did_type'], did_scope, dids['name']))
+                    new_item['type'] = dids['did_type'].upper()
+                    new_item['name'] = dids['name']
+                    new_item['did'] = '%s:%s' % (did_scope, dids['name'])
+                    resolved_items.append(new_item)
+            else:
+                new_item['type'] = self.client.get_did(did_scope, did_name)['type'].upper()
+                new_item['name'] = did_name
+                resolved_items.append(new_item)
+
+        # this list will have one dict for each file to download
+        file_items = []
+
+        # get replicas for every file of the given dids
+        logger.debug('%d DIDs after processing input' % len(resolved_items))
+        for item in resolved_items:
+            did_scope = item['scope']
+            did_name = item['name']
+            did_str = item['did']
+
+            logger.debug('Processing: %s' % item)
+
+            # since we are using metalink we need to explicitly
+            # give all schemes (probably due to a bad server site implementation)
+            force_scheme = item.get('force_scheme')
+            if force_scheme:
+                schemes = force_scheme if isinstance(force_scheme, list) else [force_scheme]
+            else:
+                schemes = ['davs', 'gsiftp', 'https', 'root', 'srm', 'file']
```

Comment:
As just discussed, so that this is documented here too. This hardcoding of schemes is bad. I understand that this is how the `list_replicas` with `metalink=True` requires it, but it really shouldn't. The default behaviour of list_replicas should be if schemes = [] is given, all protocols should be supplied. That's the only way how we can make proper protocol failover work. Due to release considerations, I am in favour of merging this, but we should address this eventually. See #1262